File: adapter_example_go/adapter.go
package main

// Adapter wraps an Adaptee and makes it compatible with Target.
type Adapter struct {
	adaptee Adaptee
}

func NewAdapter(adaptee Adaptee) *Adapter {
	return &Adapter{adaptee: adaptee}
}

func (a *Adapter) Request() string {
	translated := reverse(a.adaptee.SpecificRequest())
	return "Adapter: (TRANSLATED) " + translated
}

func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

File: adapter_example/adapter.py
from typing import Final

from target import Target
from adaptee import Adaptee


class Adapter(Target):
    def __init__(self, adaptee: Adaptee) -> None:
        self._adaptee: Final[Adaptee] = adaptee

    def request(self) -> str:
        translated = self._adaptee.specific_request()[::-1]
        return f"Adapter: (TRANSLATED) {translated}"